### ISSUE TITLE
Overlapping strides not supported by cublas

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#30596]
https://twitter.com/ID_AA_Carmack/status/1201309356089921537

Related https://github.com/pytorch/pytorch/issues/29984 but generally broadcasting is easier to support than arbitrary overlapping strides.

**Comment by ngimel** [comment]
BLAS standard dictates that arguments to blas<T>gemm call are non-overlapping https://www.math.utah.edu/software/lapack/lapack-blas/sgemm.html (look for descriptions of LDA and LDB parameters), similar for blas<T>ger and blas<T>gemv from #29984, incx and incy must be nonzero according to blas. So it's not a cublas bug, the question here is do we in pytorch want workarounds for those cases, or just better error messages.
An easy but not necessarily performant workaround is copying inputs to a contiguous tensor, but note, we are not doing this for convolutions, and instead return cudnn error with a comment that possibly the inputs are non-contiguous.

**Comment by ezyang** [comment]
Yeah, I'm not sure how to deal here. Torn between "make it obvious when you're doing something that would be very slow" and "make it always work, no matter what, even if it's slow in some regimes"
